Chap_API_Storage.tex

```diff
@@ -0,0 +1,225 @@
+%
+\declareconstitemNEW{PMIX_STORAGE_MEDIUM_PMEM}
+The storage system uses persistent memory.
+%
+\declareconstitemNEW{PMIX_STORAGE_MEDIUM_RAM}
+The storage system is volatile (e.g., tmpfs).
+%
+\end{constantdesc}
+
+\declarestruct{pmix_storage_locality_t}
+
+The \refstruct{pmix_storage_locality_t} structure is an enumerated type for specifying the locality of storage system resources. The locality is expressed relative to the process querying the attribute.
+
+\begin{constantdesc}
+%
+\declareconstitemNEW{PMIX_STORAGE_LOCALITY_NODE}
+The storage system resources are node-local.
+%
+\declareconstitemNEW{PMIX_STORAGE_LOCALITY_NETWORK}
+The storage system resources are network-local (e.g., on the same switch).
+%
+\declareconstitemNEW{PMIX_STORAGE_LOCALITY_SITE}
+The storage system resources are site-local (i.e., in the same datacenter).
+%
+\declareconstitemNEW{PMIX_STORAGE_LOCALITY_WAN}
+The storage system resources are on the WAN (i.e., not site-local).
+%
+\end{constantdesc}
+
+
+\declarestruct{pmix_storage_access_type_t}
+
+The \refstruct{pmix_storage_access_type_t} structure is an enumerated type for specifying storage system access types.
+
+\begin{constantdesc}
+%
+\declareconstitemNEW{PMIX_STORAGE_ACCESS_RD}
+Provide information on storage system read operations.
+%
+\declareconstitemNEW{PMIX_STORAGE_ACCESS_WR}
+Provide information on storage system write operations.
+%
+\declareconstitemNEW{PMIX_STORAGE_ACCESS_RDWR}
+Provide information on storage system read and write operations.
+%
+\end{constantdesc}
+
+
+%%%%%%%%%%%
+\section{Storage support attributes}
+\label{api:struct:attributes:pstrg}
+
+The following attributes may be returned in response to queries (e.g., \refapi{PMIx_Get} or \refapi{PMIx_Query_info}) made by processes or tools.
+
+%
+\declareAttributeNEW{PMIX_STORAGE_ID}{"pmix.strg.id"}{char*}{
+An identifier for the storage system (e.g., lustre-fs1, daos-oss1, home-fs)
+}
+
+%
+\declareAttributeNEW{PMIX_STORAGE_PATH}{"pmix.strg.path"}{char*}{
+Mount point path for the storage system (valid only for file-based storage systems)
+}
+
+%
+\declareAttributeNEW{PMIX_STORAGE_TYPE}{"pmix.strg.type"}{pmix_storage_type_t}{
+Type of storage system (i.e., Lustre, GPFS, DAOS, ext4, tmpfs)
+}
+
+%
+\declareAttributeNEW{PMIX_STORAGE_MEDIUM}{"pmix.strg.medium"}{pmix_storage_medium_t}{
+Types of storage mediums utilized by the storage system (e.g., SSDs, HDDs, tape)
+}
+
+%
+\declareAttributeNEW{PMIX_STORAGE_LOCALITY}{"pmix.strg.local"}{pmix_storage_locality_t}{
+Locality of storage resources provided by the storage system (e.g., node-local or network-local)
+}
+
+%
+\declareAttributeNEW{PMIX_QUERY_STORAGE_LIST}{"pmix.strg.list"}{char*}{
+Comma-delimited list of storage identifiers (i.e., \refattr{PMIX_STORAGE_ID} types) for available storage systems
+}
+
+%
+\declareAttributeNEW{PMIX_STORAGE_CAPACITY_LIMIT}{"pmix.strg.caplim"}{uint64_t}{
+Overall limit on capacity (in Megabytes[base2]) for the storage system
```

Comment:
Storage capacity is typically given in base 10 (at least for single drives).  Would it be worthwhile to export this as just bytes to avoid the base 2 vs base 10 issue?

Reply:
I think we decided against bytes mostly just to future-proof for exabyte storage systems (uint64_t can only represent up to 16 exabytes of storage, which is massive but we already expect OLCF Frontier to have 1 exabyte of capacity).

We chose base2 just because most of us felt that was more of the widely-used standard for storage system and operating system _software_ -- I know that's at odds with how storage hardware and network hardware/software is typically specified, but it does seem to be pretty commonplace. Hopefully being explicit in the definition is sufficient, but do agree that this inconsistency can be confusing in general.

Reply:
We could do it as a float instead of an int...

Reply:
Yes, that would certainly buy us some space. Somewhat unconventional in the sense that bytes would never have a fractional value, but maybe it's better than the alternatives (not enough numerical range or having to deal with base 2 vs base 10 confusion)?